presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -134,9 +134,22 @@ public Visitor(Metadata metadata, TypeAnalyzer typeAnalyzer, Session session, Ty
         public Expression rewriteComparisonExpression(ComparisonExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
         {
             ComparisonExpression expression = (ComparisonExpression) treeRewriter.defaultRewrite((Expression) node, null);
+            expression = normalizeExpression(expression);
```

Comment:
I think the idea was that this is handled by `CanonicalizeExpressions`, so it's not needed here.
What is the benefit of adding this here? (Is it related to tpcds layouts?)

Reply:
I talked to Praveen offline about this. The issue is that some expressions that are constant may have intervening casts because there’s no way to represent a typed literal in the AST. For instance, cast(x as decimal(5, 2) = cast(3 as decimal(5, 2))

Reply:
Yeah. `CanonicalizeExpressions` normalize the expression if the constant is of type `Literal` but here we cast them so it was failing I guess. Not sure if this patch handles them.

Reply:
Should we extend `CanonicalizeExpressions` to cover "almost literal" (whatever `LiteralEncoder` produces) as well?

Reply:
Yes. There are two ways of doing this: 
1. Use ExpressionInterpreter.optimize() to see if either side folds to a constant.
2. Try to match Cast(Literal) directly.

I'm a little concerned about 1) since ExpressionInterpreter.optimize() is not lightweight, especially for something that might be evaluated multiple times for all expressions in the plan. I suggest we go with 2) for now and add a comment explaining why we have to do that.

@Praveen2112, can you submit that as a separate PR? I'd like to get that in the next version if possible, as the UnwrapCast optimization will benefit from it. Also, make sure to add some tests.

Reply:
Also, this PR is ready to go, but will likely be affected by the same issue: https://github.com/prestosql/presto/pull/689. So, another reason to do it independently from this PR.

Reply:
@martint I'll submit a separate PR for the second approach